src/main/java/org/openjdk/shenandoah/ShenandoahVisualizer.java

```diff
@@ -755,14 +796,33 @@ public synchronized void renderGraph(Graphics g) {
                 g.setColor(Colors.LIVE_CSET);
                 g.drawRect(x, (int) Math.round(startRaw - s.collectionSet() * stepY), 1, 1);
 
-                // Draw this in the lower band.
-                final int smooth = Math.min(10, i + 1);
-                final int mult = 50;
+                g.setColor(Color.GRAY);
+                g.drawString("OM", bandWidth + 10, bandHeight + pad + 20);
+                g.drawString("M", bandWidth + 10, bandHeight + phaseHeight + pad + 20);
+                g.drawString("E", bandWidth + 10, bandHeight + 2*phaseHeight + pad + 20);
+                g.drawString("UR", bandWidth + 10, bandHeight + 3*phaseHeight + pad + 20);
+
+                Graphics2D g2 = (Graphics2D) g;
+                g2.setStroke(new BasicStroke(2));
+                g2.drawLine(0, bandHeight + 5, 0, bandHeight + pad - 5);
+                g2.drawLine(bandWidth / 4, bandHeight + 5, bandWidth / 4, bandHeight + pad - 5);
+                g2.drawLine(bandWidth / 2, bandHeight + 5, bandWidth / 2, bandHeight + pad - 5);
+                g2.drawLine(bandWidth * 3 / 4, bandHeight + 5, bandWidth * 3 / 4, bandHeight + pad - 5);
+
+                g.drawString("-" + Long.toString(lastSnapshots.get(lastSnapshots.size() - 1). time() - lastSnapshots.get(0).time()) + " ms", 3, bandHeight + 20);
+
+                if (lastSnapshots.size() > (graphWidth - 50) / STEP_X / 4 ) {
+                    g.drawString("-" + Long.toString(lastSnapshots.get(lastSnapshots.size() - 1). time() - lastSnapshots.get(oneFourthIndex).time()) + " ms", bandWidth / 4 + 3, bandHeight + 20);
```

Comment:
There's a lot going on here. I see the `(graphWidth - 50) / STEP_X` duplicated a few times. I think it would be clearer if we had a method like `renderTimestampLabel`.

Reply:
Okay I'll try to create a method `renderTimestampLabel` and includes all the part that it generates timestamp.